### PR TITLE
[WIP] Resource collection create from iterator

### PR DESCRIPTION
**Pull request by lsmith77**

### CODE REVIEW

File: lib/ResourceCollection.php [hunk 1]
```diff
@@ -37,11 +37,43 @@ public function __construct(HttpClientInterface $client, array $collection = arr
     }
 
     /**
+     * @param HttpClientInterface $client
+     * @param \Iterator $collection
+     *
+     * @return ResourceCollection
+     */
+    static public function createFromIterator(HttpClientInterface $client, \Iterator $collection)
```
Comment (pokap) [inline lib/ResourceCollection.php]: `static` must be declared after the visibility ;)

File: lib/EntryPoint.php [hunk 2]
```diff
@@ -38,7 +39,7 @@ public function __construct($url, HttpClientInterface $client, array $headers =
         $this->client  = $client;
         $this->headers = $headers;
 
-        $this->resource = false;
```
Comment (lsmith77) [inline lib/EntryPoint.php:41]: from a quick look it didn't seem necessary to initialize it to false here and so it seems to just confuse users into thinking that a boolean is a sensible value. imho `null|Resource` are sufficient

File: lib/ResourceCollection.php [hunk 3]
```diff
@@ -37,11 +37,43 @@ public function __construct(HttpClientInterface $client, array $collection = arr
...
+
+        return $col;
+    }
+
+    /**
+     * @param array|null|Resource $data
+     *
+     * @return Resource
+     */
+    protected function createResource($data)
```
Comment (lsmith77) [inline lib/ResourceCollection.php:55]: this method is a bit more messy than I intended ..

File: lib/ResourceCollection.php [hunk 4]
```diff
@@ -28,7 +28,7 @@ public function __construct(HttpClientInterface $client, array $collection = arr
         $this->client     = $client;
 
         foreach ($collection as $pos => $data) {
-            $collection[$pos] = Resource::create($client, $data);
+            $collection[$pos] = $this->createResource($data);
```
Comment (lsmith77) [inline lib/ResourceCollection.php]: one could also make this conversion lazy

File: lib/ResourceCollection.php [hunk 5]
```diff
@@ -97,7 +140,21 @@ public function offsetExists($offset)
      */
     public function offsetGet($offset)
     {
-        return $this->collection[$offset];
+        if (!$this->iterator instanceof \ArrayAccess) {
+            throw new \RuntimeException('Operation not allowed');
+        }
+
+        $resource = $this->iterator->offsetGet($offset);
```
Comment (lsmith77) [inline lib/ResourceCollection.php:147]: this also fixes a minor bug, in that `offsetGet` should return null when an undefined offset is passed in

**Comment by lsmith77** [comment]
tests are passing .. but I guess new tests should be added too.

in general I am not quite sure if this is really the way to go.
however I think the current `collection` property wasn't needed. also with this approach the Resource instances are created lazy. then again, not sure what the normal usage pattern is .. always read all elements or just read specific ones.

**Comment by rande** [comment]
This code looks a bit more complex, and I don't see any real benefit.

**Comment by lsmith77** [comment]
it would allow dealing with larger results as one could for example have a streaming iterator. furthermore even without such a streaming iterator, at least the entire collection does not have to be mapped to `Resource` instance if one only uses a few of the items in the collection.

but yeah its quite a bit more complex.

if this does not get merged, I would still recommend to drop the `collection` property and I would handle the case when an offset is not defined in `offsetGet()`

**Comment by lsmith77** [comment]
see https://github.com/ekino/php-hal-client/pull/7

**Comment by lsmith77** [comment]
so my argument convinced you .. or you merged the wrong PR? :)

**Comment by rande** [comment]
@lsmith77 you got the right arguments. That is why I ask for review ;)

**Pull request merged**